azext_iot/iothub/providers/state.py

```diff
@@ -0,0 +1,310 @@
+                               secondary_thumbprint=stp, status=status, status_reason=status_reason)
+
+        else:
+            logger.error("Authorization type for device '{0}' not recognized.".format(device_id))
+
+    def upload_module_identity(self, identity: dict):
+
+        device_id = identity["device_id"]
+        module_id = identity["module_id"]
+        auth_type = identity["authentication"]["type"]
+
+        if auth_type == DeviceAuthApiType.sas.value:
+            pk = identity["authentication"]["symmetricKey"]["primaryKey"]
+            sk = identity["authentication"]["symmetricKey"]["secondaryKey"]
+
+            _iot_device_module_create(self.target, device_id, module_id, primary_key=pk, secondary_key=sk)
+
+        elif auth_type == DeviceAuthApiType.selfSigned.value:
+            ptp = identity["authentication"]["x509Thumbprint"]["primaryThumbprint"]
+            stp = identity["authentication"]["x509Thumbprint"]["secondaryThumbprint"]
+
+            _iot_device_module_create(self.target, device_id, module_id, DeviceAuthType.x509_thumbprint.value,
+                                      primary_thumbprint=ptp, secondary_thumbprint=stp)
+
+        elif auth_type == DeviceAuthApiType.certificateAuthority.value:
+            _iot_device_module_create(self.target, device_id, module_id, DeviceAuthType.x509_ca.value)
+
+        else:
+            logger.error("Authorization type for module '{0}' in device '{1}' not recognized.".format(module_id, device_id))
+
+    def delete_all_configs(self):
+        configs = _iot_hub_configuration_list(self.hub_name, self.target)
+        for i in tqdm(range(len(configs)), desc="Deleting configurations and edge deployments from destination hub"):
+            c = configs[i]
+            _iot_hub_configuration_delete(self.target, config_id=c["id"])
+
+    def delete_all_devices(self):
+        identities = _iot_device_list(self.hub_name, self.target, top=-1)
+        for i in tqdm(range(len(identities)), desc="Deleting identities from destination hub"):
+            id = identities[i]
+            _iot_device_delete(self.target, id["deviceId"])
+
+    def upload_state(self, filename: str, replace: Optional[bool] = None):
+        '''
+        Uses device info from file to recreate the devices
+        '''
+
+        if replace:
+            self.delete_all_configs()
+            self.delete_all_devices()
+
+        with open(filename, 'r', encoding='utf-8') as f:
+            hub_state = json.load(f)
+
+        self.upload_hub_from_dict(hub_state)
+
+        logger.info("Uploaded state from '{}' to IoT Hub '{}'".format(filename, self.hub_name))
+
+    def migrate_devices(
+        self,
+        orig_hub: Optional[str] = None,
+        orig_rg: Optional[str] = None,
+        orig_hub_login: Optional[str] = None,
+        replace: Optional[bool] = False,
+    ):
+
+        if replace:
```

Comment:
do the replace after you get the state for the original hub - think of what happens if the two hubs are the same